homeassistant/components/rainmachine/__init__.py

```diff
@@ -0,0 +1,240 @@
+    {
+        DOMAIN: vol.Schema({
+            vol.Required(CONF_IP_ADDRESS): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+            vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+            vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+            vol.Optional(CONF_BINARY_SENSORS): BINARY_SENSOR_SCHEMA,
+            vol.Optional(CONF_SENSORS): SENSOR_SCHEMA,
+            vol.Optional(CONF_SWITCHES): SWITCH_SCHEMA,
+        })
+    },
+    extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the RainMachine component."""
+    from regenmaschine import Authenticator, Client
+    from regenmaschine.exceptions import RainMachineError
+
+    conf = config[DOMAIN]
+    ip_address = conf[CONF_IP_ADDRESS]
+    password = conf[CONF_PASSWORD]
+    port = conf[CONF_PORT]
+    ssl = conf[CONF_SSL]
+
+    try:
+        auth = Authenticator.create_local(
+            ip_address, password, port=port, https=ssl)
+        rainmachine = RainMachine(hass, Client(auth))
+        rainmachine.update()
+        hass.data[DATA_RAINMACHINE] = rainmachine
+    except RainMachineError as exc:
+        _LOGGER.error('An error occurred: %s', str(exc))
+        hass.components.persistent_notification.create(
+            'Error: {0}<br />'
+            'You will need to restart hass after fixing.'
+            ''.format(exc),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    binary_sensor_config = conf.get(CONF_BINARY_SENSORS, {})
+    discovery.load_platform(
+        hass, 'binary_sensor', DOMAIN, binary_sensor_config, config)
+
+    sensor_config = conf.get(CONF_SENSORS, {})
+    discovery.load_platform(
+        hass, 'sensor', DOMAIN, sensor_config, config)
+
+    switch_config = conf.get(CONF_SWITCHES, {})
+    discovery.load_platform(hass, 'switch', DOMAIN, switch_config, config)
+
+    def refresh(event_time):
+        """Refresh RainMachine data."""
+        _LOGGER.debug('Updating RainMachine data')
+        hass.data[DATA_RAINMACHINE].update()
+        dispatcher_send(hass, DATA_UPDATE_TOPIC)
+
+    track_time_interval(hass, refresh, DEFAULT_SCAN_INTERVAL)
+
+    def start_program(service):
+        """Start a particular program."""
+        rainmachine.client.programs.start(service.data[CONF_PROGRAM_ID])
+
+    def start_zone(service):
+        """Start a particular zone for a certain amount of time."""
+        rainmachine.client.zones.start(
+            service.data[CONF_ZONE_ID], service.data[CONF_ZONE_RUN_TIME])
+
+    def stop_all(service):
+        """Stop all watering."""
+        rainmachine.client.watering.stop_all()
+
+    def stop_program(service):
+        """Stop a program."""
+        rainmachine.client.programs.stop(service.data[CONF_PROGRAM_ID])
+
+    def stop_zone(service):
+        """Stop a zone."""
+        rainmachine.client.zones.stop(service.data[CONF_ZONE_ID])
+
+    hass.services.register(
```

Comment:
Do a for loop here too. Iterate over service name, service handler and config schema. Use empty dict when there's no specific schema.